apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/VolatileResourceManager.cs

```diff
@@ -27,28 +27,48 @@ internal sealed class VolatileResourceManager : ISinglePhaseNotification, ISpann
         private readonly SpannerConnection _spannerConnection;
         private readonly TimestampBound _timestampBound;
         private readonly TransactionId _transactionId;
-        private SpannerTransaction _transaction;
+        private Lazy<Task<SpannerTransaction>> _transaction;
+        private bool _hasExecutedDml;
 
         internal VolatileResourceManager(SpannerConnection spannerConnection, TimestampBound timestampBound, TransactionId transactionId)
         {
             _spannerConnection = spannerConnection;
             _timestampBound = timestampBound;
+            _transaction = new Lazy<Task<SpannerTransaction>>(CreateTransactionAsync, LazyThreadSafetyMode.ExecutionAndPublication);
             _transactionId = transactionId;
         }
 
+        private SpannerTransaction SpannerTransaction => SpannerTransactionTask.Result;
+
+        private Task<SpannerTransaction> SpannerTransactionTask => _transaction == null
+            ? throw new ObjectDisposedException("The transaction has already been disposed")
+            : _transaction.Value;
+
+        private bool IsTransactionCreated => _transaction != null && _transaction.IsValueCreated;
+
+        private bool HasExecutedDmlOrMutations => IsTransactionCreated && (_hasExecutedDml || SpannerTransaction.HasMutations);
+
         private Logger Logger => _spannerConnection.Logger;
 
+        private Task<SpannerTransaction> CreateTransactionAsync()
+        {
+            return _timestampBound != null ? _spannerConnection.BeginReadOnlyTransactionAsync(_timestampBound)
+                : _transactionId != null ? Task.FromResult(_spannerConnection.BeginReadOnlyTransaction(_transactionId))
+                : _spannerConnection.BeginTransactionAsync();
+        }
+
         public void Dispose()
         {
             try
             {
-                _transaction?.Dispose();
-                // Protect against multiple disposals
+                if (IsTransactionCreated)
```

Comment:
Actually I think you want here `_transaction is null` only, without caring about the value having been created or not.

Reply:
No, in this case we do care, because it could be that the `VolatileResourceManager` is being disposed without the `SpannerTransaction` ever having been initialized. `_transaction` is set to a [non-null value in the constructor](https://github.com/googleapis/google-cloud-dotnet/blob/5a20380ab74dbe7570cb2d3cb6480d32c05bad40/apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/VolatileResourceManager.cs#L37). The underlying `SpannerTransaction` is however only created if it is actually needed. So if `Dispose()` would be called before the transaction has been used, it would unnecessarily first create a new `SpannerTransaction` and then dispose it right afterwards.

Reply:
Yep, you are right.